scripts/image_downloader.py

```diff
@@ -0,0 +1,122 @@
+"""
+    This is a script that takes data from the CSVs, download and stores the images.
+    It assumes the csv has the following columns: name, thumbnail link, date, img link, subcategory id, subcategory
+
+    The final dataset is stored in the final_filename{i} files according to CIFAR semantics:
+
+    The archive contains the files dataset1, dataset2, ... Each of these files is a Python "pickled" object produced with Pickle. 
+    Here is a python3 routine which will open such a file and return a dictionary:
+
+    def unpickle(file):
+        import pickle
+        with open(file, 'rb') as fo:
+            dict = pickle.load(fo, encoding='bytes')
+        return dict
+    
+    Loaded in this way, each of the batch files contains a dictionary with the following elements:
+    data -- a 10000x7500 numpy array of uint8s. Each row of the array stores a 50x50 colour image. The first 2500 entries contain the red channel values, the next 2500 the green, and the final 2500 the blue.
+    The image is stored in row-major order, so that the first 50 entries of the array are the red channel values of the first row of the image.
+    labels -- a list of 10000 numbers in the range 0-20. The number at index i indicates the label of the ith image in the array data.
+
+    Dependencies: Pillow, Pickle, numpy, requests
+"""
+
+from os import listdir
+import csv
+from enum import Enum
+import requests
+import numpy as np
+import pickle
+import io
+import PIL.Image as Image
+import PIL.ImageOps as ImageOps
+import threading
+
+class Tags(Enum):
+    NAME = 0
+    THUMBNAIL = 1
+    DATE = 2
+    IMG = 3
+    SUB_ID = 4
+    SUB = 5
+
+def download_file_images(file, data, labels, lock: threading.RLock, image_size):
+    global collected_samples
+    csv_filename = DATASET_DIR + file
+    with open(csv_filename, 'r', newline='') as csvfile:
+        reader = csv.reader(csvfile, delimiter=',')
+        #read each line
+        for index, row in enumerate(reader):
+            #take one in ten
+            if index % (100 / percentage2download) == 0:
+                date = row[Tags.DATE.value]
+                #check date
+                if not date or int(date.split("-")[0]) > date_accepted:
+                    thumb_url = row[Tags.THUMBNAIL.value]
+                    
+                    answer = requests.get(thumb_url, headers=headers) #download thumbnail
+                    if answer.status_code != 404:
+                        image = answer.content
+                        #resize image
+                        image = Image.open(io.BytesIO(image))
+                        image = ImageOps.pad(image, image_size)
```

Comment:
Can we remove this? So that we only crawl once. The processing is done later before training.

Reply:
Fixed now